### ISSUE TITLE
Questions about 3D-MoLM

### ISSUE THREAD

**Issue by icycookies** [lsh0520/3D-MoLM#8]
Thanks again for sharing this innovative work! However, I encountered several problems during reproduction:

- In `model/blip2qformer.py` L317, the encoder attention mask is directly derived from Uni-mol. However, I observed that this mask is contrary with the format of Transformers. As a result, the Q-Former extracts representations based on the padding representations of Uni-Mol, which seems to integrate certain global information of the molecule.
- The `stage1-ft.ckpt` works fine with cross-modal retrieval. However, when I try the `stage1.ckpt` by running `python stage1.py --filename stage1_ft --mode eval --max_epochs 10 --warmup_steps 200 --lm --init_checkpoint all_checkpoints/stage1.ckpt` on the test set of PubChem,  I got the following results:

<img width="658" alt="屏幕快照 2024-04-09 下午9 17 50" src="https://github.com/lsh0520/3D-MoLM/assets/18654907/0c8adba0-17ca-49b6-a061-50295a3365c0">

 Since the Stage1 checkpoint is pre-trained on the pre-training set of PubChem, I wonder if this sub-optimal result is the case for 3D-MoLM.

Looking forward to your reply!

**Comment by lsh0520** [comment]
(1) @acharkq Zhiyuan, can you help with the first question? Thanks.
(2) No. I have uploaded the training log with evaluation results of stage 1 pretrained model to huggingface, please check the performance of it. The results in the picture doesn't make sense, which could be (I guess): (1) the environt problem, can you check if your environment is the same with the requirements.txt, (2) I wrongly uploaed a corrupted ckpt. I will look into this later this week and make sure you can reproduce the results in our paper with the ckpt and scripts.

**Comment by acharkq** [comment]
Thanks @icycookies for raising the issue about uni-mol's encoder mask. I have checked and it is indeed contrary to the format of huggingface's transformers. Luckily, the padding tokens can still capture some global information, thus the model still works. I will discuss with sihang and see how we will deal with bug.

**Comment by pansanity666** [comment]
Hi, 

The padding tokens are fixed with zeros with no gradients in the nn.Embedding of unimol encoder. And the padding part is masked with -inf in the attention. 

My question is: **is the padding token really capturing certain global information? or just no information about 3D at all?**

Best,

**Comment by acharkq** [comment]
Hi, 

Thanks for your interest. It is clear that it works somehow. Otherwise the stage 1 will have 0 performance

**Comment by pansanity666** [comment]
Hi, 

I checked the attn_bias added to the self-attention of unimol encoder:

<img width="288" alt="image" src="https://github.com/lsh0520/3D-MoLM/assets/47111102/135e86e7-10b4-4496-84b0-c2da75325237">

It seems that the attn_bias is actually not properly processed. Therefore, the last few rows (correspond to padding token) still get attention weight towards atom tokens, i.e., 17.8750. 

I think this is the reason for the global info in padding tokens. 

Best,

**Comment by acharkq** [comment]
Thanks for the efforts on proving this
